Blogpost on multilingual documentation (#796)
ropensci/roweb3: content/blog/2024-07-09-multilingual-documentation-in-r/index.es.Rmd

---
slug: multilingual-documentation-wg
title: Documentación multilingüe en paquetes de R
author:
- Elio Campitelli

> date: '2024-07-09'

Reviewer: ```suggestion
editor:
  - Steffi LaZerte
  - Yanina Bellini Saibene
date: '2024-07-09'
```

---

> slug: 2024-07-09-multilingual-documentation

Reviewer 2: ```suggestion
slug: multilingual-documentation
```

title: Documentación multilingüe en paquetes de R
author:
- Elio Campitelli
date: '2024-07-09'
tags:
- R
- packages
- multilingual
- documentation
- community
description: Documentación multilingüe en un paquete R cercano

> social: "La documentación multilingüe llega a un paquete de R cercano @eliocamp@mastodon.social"

Reviewer: ```suggestion
social: "La documentación multilingüe de paquetes de R al alcance de tu mano @eliocamp@mastodon.social"
```

  }
)

# knitr hook to use Hugo highlighting options
knitr::knit_hooks$set(
  source = function(x, options) {
    hlopts <- options$hlopts
    paste0(
      "```r ",
      if (!is.null(hlopts)) {
        paste0("{",
               glue::glue_collapse(
                 glue::glue('{names(hlopts)}={hlopts}'),
                 sep = ","
               ), "}"
        )
      },
      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
    )
  }
)

hook_output <- knitr::knit_hooks$get("output")
knitr::knit_hooks$set(output = function(x, options) {
  lines <- options$output.lines
  if (is.null(lines)) {
    return(hook_output(x, options))  # pass to default hook
  }
  x <- unlist(strsplit(x, "\n"))
  
  if (length(x) > lines) {
    more <- paste0("... (", length(x) - lines, " more lines)")
    # truncate the output, but add ....
    x <- c(head(x, lines), more)
  }
  
  # paste these lines together
  x <- paste(c(x, ""), collapse = "\n")
  hook_output(x, options)
})

help <- function(...) {
  outfile <- tempfile(fileext = ".txt")
  tools::Rd2txt(utils:::.getHelpFile(utils::help(...)), outfile)
  text <- readLines(outfile)
  text <- gsub("_\b", "", text)
  cat(text, sep = "\n")
}
````

> Nos complace anunciar el nuevo [Grupo de Trabajo Multilingüe del Consorcio R](https://github.com/RConsortium/multilingual-documentation-wg).

Reviewer: ```suggestion
Nos complace anunciar el nuevo [Grupo de Trabajo Multilingüe del R Consortium](https://github.com/RConsortium/multilingual-documentation-wg).
```

> Este Grupo de Trabajo surgió tras los debates mantenidos durante la [Proyecto R Sprint 2023](https://contributor.r-project.org/r-project-sprint-2023/) y supervisará la implementación del soporte de documentación multilingüe en R y organizará los esfuerzos de traducción de la comunidad.

Reviewer: ```suggestion
Este Grupo de Trabajo surgió tras los debates mantenidos durante el [R Project Sprint 2023](https://contributor.r-project.org/r-project-sprint-2023/) y supervisará la implementación del soporte de documentación multilingüe en R y organizará los esfuerzos de traducción de la comunidad.
```

Nuestro primer proyecto es el (experimental) [**rhelpi18n**](https://github.com/eliocamp/rhelpi18n) que añade documentación multilingüe.

> ¿Desea leer la documentación de `base::mean()` (parcialmente) en español?

Reviewer: ```suggestion
¿Quieres leer la documentación de `base::mean()` (parcialmente) en español?
```

> Instale la versión experimental [rhelpi18n](https://github.com/eliocamp/rhelpi18n) y el módulo de traducción de ejemplo base.es:

Reviewer: ```suggestion
Instala la versión experimental [rhelpi18n](https://github.com/eliocamp/rhelpi18n) y el módulo de traducción de ejemplo base.es:
```

```{r, eval=FALSE}
# install.packages("pak")
pak::pak(c("eliocamp/rhelpi18n", "eliocamp/base.es"))
```

Y luego ejecuta este código:

```{r, output.lines=30, warning=FALSE}
library(rhelpi18n)
Sys.setenv(LANGUAGE = "es")
help(mean) 
```

¡Bien!
Esto debería funcionar en cualquier consola R o IDE.

## ¿Por qué harías esto?

Aunque el inglés es la lengua internacional de facto, existen [ventajas](https://ropensci.org/multilingual-publishing/) de incluir recursos en diferentes idiomas.
Al reducir las barreras lingüísticas, acabamos teniendo una comunidad más grande, fuerte y creativa, con más ideas y recursos.

> Esta hegemonía del inglés se refleja en R por el uso del inglés en su documentación, como las páginas de manual, los nombres de las funciones y los nombres de los argumentos (¿por qué usar `mean()` en lugar de `Mittelwert()`?)

Reviewer: ```suggestion
Esta hegemonía del inglés se refleja en R por el uso del inglés en su documentación, como las páginas de ayuda, los nombres de las funciones y los nombres de los argumentos (¿por qué usar `mean()` en lugar de `Mittelwert()`?)
```

.

> Documentación del paquete *puede* puede escribirse en cualquier idioma, y CRAN admite documentación en otros idiomas utilizando el campo "Language" (Idioma) [^1] .

Reviewer: ```suggestion
La documentación de los paquetes *puede* escribirse en cualquier idioma, y CRAN admite documentación en otros idiomas utilizando el campo "Language" (Idioma) [^1] .
```

> Pero aún así, la gran mayoría de los paquetes están documentados en inglés y el campo [pequeño número de paquetes documentados en otros idi](https://cderv.rbind.io/2018/03/11/non-english-pkg-in-cran/) parecen estar adaptados a su público objetivo.

Reviewer: ```suggestion
Pero aún así, la gran mayoría de los paquetes están documentados en inglés y el [pequeño número de paquetes documentados en otros idimas](https://cderv.rbind.io/2018/03/11/non-english-pkg-in-cran/) se ajustan aparentemente a su público objetivo.
```

> Por ejemplo, los paquetes [labstatR](https://cran.r-project.org/web/packages/labstatR/index.html) sirve de complemento al libro italiano "Laboratorio Di Statistica Con R" y está parcialmente documentado en italiano (utiliza nombres de funciones y argumentos en inglés).

Reviewer: ```suggestion
Por ejemplo, el paquete [labstatR](https://cran.r-project.org/web/packages/labstatR/index.html) sirve de complemento al libro italiano "Laboratorio Di Statistica Con R" y está parcialmente documentado en italiano (utiliza nombres de funciones y argumentos en inglés).
```

Del mismo modo, [chilemapas](https://cran.r-project.org/web/packages/chilemapas/chilemapas.pdf) proporciona mapas simplificados para Chile, con documentación completa en español, incluidos los nombres de las funciones.
rOpenSci [censo2017](https://docs.ropensci.org/censo2017/) por [Mauricio Vargas](/author/pachá-aka-mauricio-vargas-sepúlveda/) también está totalmente documentado en español.

> [^1]: La cita real es "Se puede utilizar un campo 'Idioma' para indicar si la documentación del paquete no está en inglés", lo que indica que el inglés sigue siendo el supuesto por defecto.

Reviewer: ```suggestion
[^1]: La cita real es "Se puede utilizar el campo 'Language' para indicar si la documentación del paquete no está en inglés", lo que indica que el inglés sigue siendo el supuesto por defecto.
```

Se trata de un problema difícil que no tiene fácil solución.
Surge periódicamente en el Slack de rOpenSci, así como en otros espacios de R, como Stack Overflow.
El desarrollador del [utilsIPEA](https://cran.r-project.org/web/packages/utilsIPEA/index.html) expresó [la necesidad de documentación bilingüe](https://stackoverflow.com/questions/37288823/bilingual-english-and-portuguese-documentation-in-an-r-package) reconociendo que su paquete sería utilizado tanto por personas de Brasil, que podrían preferir la documentación en portugués, como por la comunidad internacional en general.
Al menos dos desarrolladores de paquetes han intentado resolver este dilema documentando su paquete en inglés y publicando una segunda versión documentada en otro idioma: [ExpDes](https://cran.r-project.org/web/packages/ExpDes/index.html) y [ExpDes.pt](https://cran.r-project.org/web/packages/ExpDes.pt/index.html) así como [orloca](https://cran.r-project.org/web/packages/orloca/index.html) y [orloca.es](https://cran.r-project.org/web/packages/orloca.es/index.html).

> En [karel](https://cloud.r-project.org/web/packages/karel/index.html) tiene dos conjuntos de funciones, uno con nombres y documentación en inglés y otro en español.

Reviewer: ```suggestion
El paquete [karel](https://cloud.r-project.org/web/packages/karel/index.html) tiene dos conjuntos de funciones, uno con nombres y documentación en inglés y otro en español.
```

> Ambas son soluciones que llevan al límite lo que el lenguaje de programación R puede hacer en términos de funcionalidad multilingüe.

Reviewer: ```suggestion
Ambas son soluciones alternativas que llevan al límite lo que el lenguaje de programación R puede hacer en términos de funcionalidad multilingüe.
```

Además, como estas soluciones no son nativas, pueden resultar difíciles de mantener y no son escalables.

> Por eso presentamos el proyecto experimental **rhelpi18n** [^2] paquete.

Reviewer: ```suggestion
Por eso presentamos el proyecto de paquete experimental **rhelpi18n** [^2].
```

> Este paquete es un banco de pruebas para implementar un soporte sin fisuras para la documentación multilingüe con el objetivo de integrarlo en el propio R una vez que madure.

Reviewer: ```suggestion
Este paquete es un banco de pruebas para implementar un soporte sin fallas para la documentación multilingüe con el objetivo de integrarlo en el propio R una vez que madure.
```

Un usuario puede cargar rhelpi18n, instalar módulos de traducción creados por la comunidad y leer la documentación de R en el idioma de su elección.

> [^2]: El nombre proviene del[numeronimo](https://en.wikipedia.org/wiki/Numeronym) de "internacionalización".

Reviewer: ```suggestion
[^2]: El nombre proviene del [neurónimo](https://en.wikipedia.org/wiki/Numeronym) de "internacionalización".
```

## De acuerdo, pero ¿cómo?

En nuestro ejemplo anterior, las cadenas traducidas se alojan en base.es, un módulo de traducción.
Se trata de un paquete normal con campos especiales en la DESCRIPTION que especifican qué paquete y versión se está traduciendo y a qué idioma.
Actualmente, las cadenas traducidas se implementan como una lista con la cadena original y la traducción de cada sección de la documentación, pero también estamos estudiando implementarlo con archivos .po tradicionales, que es el mecanismo utilizado por R para traducir mensajes y errores.

El paquete rhelpi18n modifica `utils:::.getHelpFile()` que utilizan todas las funciones de ayuda para obtener cadenas de documentación.
Añade unas pocas líneas de código que comprueban si existen módulos de traducción relevantes (como base.es) y sustituyen las cadenas originales por las traducidas sobre la marcha.

> Además de presentar la documentación traducida, rhelpi18n también proporciona herramientas para que los desarrolladores creen nuevos módulos de traducción.

Reviewer: ```suggestion
Además de presentar la documentación traducida, rhelpi18n también proporciona herramientas para que las personas que desarrollan creen nuevos módulos de traducción.
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
glue_zip <- file.path(tempdir(), "glue.zip")
glue_dir <- file.path(tempdir(), "glue-9f88e2179fc7aa6ff11214271fd61bba169f925c")
dir.create(glue_dir, showWarnings = FALSE, recursive = TRUE)
glue_url <- "https://github.com/tidyverse/glue/archive/9f88e2179fc7aa6ff11214271fd61bba169f925c.zip"

download.file(glue_url, glue_zip)
unzip(glue_zip, exdir = tempdir())
```

Por ejemplo, el siguiente código crea un módulo de traducción básico para el paquete glue, incluyendo todas las funcionalidades y cadenas que necesitan ser traducidas.

```{r}
dir <- file.path(tempdir(), "glue.es")
rhelpi18n::i18n_module_create(module_name = "glue.es", 

>                               package_path = glue_dir,  # Path to local copy of glue source code

Reviewer: ```suggestion
                              package_path = glue_dir,  # Camino a la copia local del codigo fuente de glue
```

                              language = "es", 
                              module_path = dir)
```

Ahora el trabajo pesado es ir al subdirectorio "translations" y traducir todas las cadenas guardadas (actualmente) como archivos YAML.
Por ejemplo, el archivo de traducción para `glue::as_glue()` es:

```{r, echo=FALSE, output.lines=20}
file.path(dir, "translations", "as_glue.yaml") |> 
  readLines() |> 
  cat(sep = "\n")
```

Una posible (y superficial) traducción sería

```{r, output.lines=30, echo=FALSE}
cat(readLines("as_glue.yaml"), sep = "\n")
```

```{r, echo=FALSE, eval=FALSE}
# file.copy("as_glue.yaml", file.path(dir, "translations", "as_glue.yaml"),
#           overwrite = TRUE)
# pak::local_install(dir)
```

Tras modificar "as\_glue.yaml" e instalar el paquete, ahora la documentación está en español .

```{r, output.lines=20}
library(glue)
help(as_glue)
```

¡Increíble!

## Próximos pasos

> Todo esto está en sus primeras fases y las cosas avanzan rápido.

Reviewer: ```suggestion
El proyecto está en sus primeras fases y las cosas avanzan rápido.
```

Aún estamos experimentando con distintas formas de almacenar las traducciones y de aplicar sustituciones de cadenas.
También tiene varias limitaciones, como que aún no traduce los encabezados de sección.
También queremos añadir una indicación clara de que el texto es una traducción y una forma cómoda de acceder rápidamente a la versión original.

> Parte del flujo de trabajo de las traducciones suele consistir en traducciones automáticas que posteriormente revisa un humano.

Reviewer: ```suggestion
Parte del flujo de trabajo de las traducciones suele consistir en traducciones automáticas que posteriormente revisa una persona.
```

> Por ello, estamos considerando la posibilidad de añadir metadatos a las cadenas para señalar los casos de traducciones automáticas que no han sido revisadas por un humano o las traducciones que podrían estar ligeramente desfasadas (¡sería una pena no mostrar una traducción porque alguien arregló una pequeña errata en el texto original y aún no había sido revisada!)

Reviewer: ```suggestion
Por ello, estamos considerando la posibilidad de añadir metadatos a las cadenas para señalar los casos de traducciones automáticas que no han sido revisadas por una persona o las traducciones que podrían estar ligeramente desfasadas (¡sería una pena no mostrar una traducción porque alguien arregló un pequeño error de tipeo en el texto original y aún no había sido revisada!)
```

> Si desea probar el paquete e informar de cualquier problema, o si tiene ideas sobre cómo mejorarlo, puede participar en el debate en la sección [temas abiertos](https://github.com/eliocamp/rhelpi18n/issues) del paquete y uniéndote al [Grupo de trabajo de documentación multilingüe](https://github.com/RConsortium/multilingual-documentation-wg).

Reviewer: ```suggestion
Si deseas probar el paquete e informar de cualquier problema, o si tienes ideas sobre cómo mejorarlo, puedes participar en el debate en la sección [_issues_ abiertos](https://github.com/eliocamp/rhelpi18n/issues) del paquete y uniéndote al [Grupo de trabajo de documentación multilingüe](https://github.com/RConsortium/multilingual-documentation-wg).
```
